File: question/neetcode150/stack/evaluate-reverse-polish-notation.py
from typing import List

class Solution:
    def evalRPN(self, a: List[str]) -> int:
        stack = []
        for n in a:
            if n.lstrip("-").isnumeric():
                stack.append(int(n))
            else:
                n2 = stack.pop()
                n1 = stack.pop()
                if n == "+":
                    n3 = n1+n2
                if n == "-":
                    n3 = n1-n2
                if n == "*":
                    n3 = n1*n2
                if n == "/":
                    if n1 * n2 < 0:
                        n3 = -(abs(n1)//abs(n2))
                    else:
                        n3 = n1//n2

                stack.append(n3)

        return stack[-1]
                

File: question/neetcode150/stack/evaluate-reverse-polish-notation.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:

    bool isNumeric(const std::string& str) {
        try {
            stoi(str);
            return true;
        } catch (...) {
            return false;
        }
    }

    int evalRPN(vector<string>& ss) {
        stack<int> s;

        for (auto x: ss){
            if(isNumeric(x)){
                s.push(stoi(x));
            }
            else{
                int y,z;
                z = s.top();
                s.pop();
                y = s.top();
                s.pop();

                if (x == "+"){
                    s.push(y+z);
                }
                if (x == "-"){
                    s.push(y-z);
                }
                if (x == "*"){
                    s.push(y*z);
                }
                if (x == "/"){
                    s.push(y/z);
                }
            }
        }

        return s.top();
    }
};